File: lib/onboarding/tdee.ts
import {
  ACTIVITY_MULTIPLIERS,
  AGGRESSION_KCAL_PER_KG,
  CARB_SPLIT_RATIOS,
} from './constants';
import type {
  ActivityLevel,
  BodyMetrics,
  CarbSplit,
  Goal,
  MacroTargets,
} from './types';

/**
 * Mifflin-St Jeor BMR formula.
 * Male:   (10 × kg) + (6.25 × cm) - (5 × age) + 5
 * Female: (10 × kg) + (6.25 × cm) - (5 × age) - 161
 * Returns raw BMR (no rounding — rounding happens in TDEE).
 */
export function calcBMR(metrics: BodyMetrics): number {
  const base =
    10 * metrics.weightKg + 6.25 * metrics.heightCm - 5 * metrics.age;
  return metrics.biologicalSex === 'male' ? base + 5 : base - 161;
}

/**
 * TDEE = BMR × activity multiplier, rounded.
 */
export function calcTDEE(bmr: number, activityLevel: ActivityLevel): number {
  return Math.round(bmr * ACTIVITY_MULTIPLIERS[activityLevel]);
}

/**
 * Compute macro grams from total calories and carb split ratio.
 * Protein: 4 cal/g, Carbs: 4 cal/g, Fat: 9 cal/g.
 */
export function calcMacroGrams(
  calories: number,
  carbSplit: CarbSplit
): MacroTargets {
  const ratio = CARB_SPLIT_RATIOS[carbSplit];
  return {
    calories: Math.round(calories),
    proteinG: Math.round((calories * ratio.protein) / 100 / 4),
    carbsG: Math.round((calories * ratio.carbs) / 100 / 4),
    fatG: Math.round((calories * ratio.fat) / 100 / 9),
  };
}

/**
 * Compute daily calorie/macro targets based on goal and aggression.
 * aggression is kg/week (0.1–0.8). deficitOverride overrides when provided.
 */
export function calcDailyTargets(
  tdee: number,
  goal: Goal,
  aggression: number | null,
  carbSplit: CarbSplit,
  deficitOverride?: number | null
): MacroTargets {
  let calories: number;

  if (goal === 'maintaining') {
    calories = tdee;
  } else {
    const safeAggression = aggression ?? 0;
    const adjustment =
      deficitOverride ?? Math.round(safeAggression * AGGRESSION_KCAL_PER_KG);
    calories = goal === 'cutting' ? tdee - adjustment : tdee + adjustment;
  }

  // NOTE: calories may go negative if deficitOverride > tdee.
  // Caller (Server Action) must clamp: Math.max(result.calories, 500) before persisting.
  return calcMacroGrams(calories, carbSplit);
}

File: apps/mobile-flutter/lib/shared/logic/tdee.dart
/// TDEE / macro maths — the app's single copy.
///
/// Vendored from web `lib/domain/onboarding/tdee.ts` +
/// `lib/domain/onboarding/constants.ts`, which remain the source of truth.
/// `test/shared/logic/tdee_test.dart` reads that TypeScript off disk and fails
/// when the numbers here drift from it.
///
/// Pure Mifflin–St Jeor: no rounding in BMR, rounding in TDEE and the macros.
/// Both the onboarding wizard and the settings profile form call this; before
/// consolidation each had its own copy with its own inlined constant tables.
library;

import '../../models/profile/onboarding.dart';

// ── Constants ────────────────────────────────────────────────────────────

const Map<ActivityLevel, double> kActivityMultipliers = {
  ActivityLevel.sedentary: 1.2,
  ActivityLevel.light: 1.375,
  ActivityLevel.moderate: 1.55,
  ActivityLevel.veryActive: 1.725,
};

/// Ratios as protein% / fat% / carbs% (must sum to 100).
class CarbSplitRatio {
  final int protein;
  final int fat;
  final int carbs;
  const CarbSplitRatio({
    required this.protein,
    required this.fat,
    required this.carbs,
  });
}

const Map<CarbSplit, CarbSplitRatio> kCarbSplitRatios = {
  CarbSplit.moderateCarb: CarbSplitRatio(protein: 30, fat: 35, carbs: 35),
  CarbSplit.lowerCarb: CarbSplitRatio(protein: 40, fat: 40, carbs: 20),
  CarbSplit.higherCarb: CarbSplitRatio(protein: 30, fat: 20, carbs: 50),
};

/// 1 kg fat ≈ 7,700 kcal → daily kcal = kg/week × 1100.
const int kAggressionKcalPerKg = 1100;

// ── Calculations ─────────────────────────────────────────────────────────

/// Mifflin–St Jeor BMR. Raw, unrounded — [calcTDEE] does the rounding.
///
/// Web passes a whole `BodyMetrics` record; this takes only the four fields the
/// formula reads, because the Dart record also carries `activityLevel` and
/// handing that to a function that ignores it reads as a bug.
double calcBMR({
  required BiologicalSex biologicalSex,
  required double weightKg,
  required int heightCm,
  required int age,
}) {
  final base = 10 * weightKg + 6.25 * heightCm - 5 * age;
  return biologicalSex == BiologicalSex.male ? base + 5 : base - 161;
}

/// TDEE = BMR × activity multiplier, rounded.
int calcTDEE(double bmr, ActivityLevel activityLevel) =>
    (bmr * kActivityMultipliers[activityLevel]!).round();

/// Rounded macro grams + calories for a calorie budget.
/// Protein 4 kcal/g, carbs 4 kcal/g, fat 9 kcal/g.
MacroTargets calcMacroGrams(double calories, CarbSplit carbSplit) {
  final ratio = kCarbSplitRatios[carbSplit]!;
  return MacroTargets(
    calories: calories.round().toDouble(),
    proteinG: (calories * ratio.protein / 100 / 4).round().toDouble(),
    carbsG: (calories * ratio.carbs / 100 / 4).round().toDouble(),
    fatG: (calories * ratio.fat / 100 / 9).round().toDouble(),
  );
}

/// Daily calorie/macro targets for a goal + pace + split.
///
/// [aggression] is kg/week (0.1–0.8); [deficitOverride] replaces the derived
/// adjustment when given. `num?` because the two pre-consolidation copies typed
/// it differently (`double?` in onboarding, `int?` in settings) and neither
/// call site passes it today.
///
/// Calories may go negative when a deficit exceeds the TDEE; the server clamps
/// `calorieTarget = max(., 500)` before persisting, mirrored by the callers.
MacroTargets calcDailyTargets(
  int tdee,
  Goal goal,
  double? aggression,
  CarbSplit carbSplit, [
  num? deficitOverride,
]) {
  double calories;

  if (goal == Goal.maintaining) {
    calories = tdee.toDouble();
  } else {
    final safeAggression = aggression ?? 0;
    final adjustment =
        deficitOverride ?? (safeAggression * kAggressionKcalPerKg).round();
    calories = goal == Goal.cutting
        ? tdee - adjustment.toDouble()
        : tdee + adjustment.toDouble();
  }

  return calcMacroGrams(calories, carbSplit);
}
